Fix response code when route not found

skrzepto opened the pull request:
return 404 accordingly when the path is invalid

skrzepto committed:
setting status code to 404 when the route is not found

skrzepto commented:
issue #45

pypingou commented:
We need to resolve the conflict, but after that, this change looks good ! :)

jeremycline commented on fmn/sse/sse_webserver.py:
```diff
@@ -29,16 +29,16 @@ class SSEServer(resource.Resource):
             return server.NOT_DONE_YET
         else:
+            request.setResponseCode(404)
             return self.invalid_request(request=request,
```
Is there a particular reason to include a JSON response body when the headers indicate 404 NOT FOUND? It's not wrong as far as I know, but it seems unnecessary.

jeremycline replied:
I noticed this sets the status reason as well so the status line comes back `HTTP/1.1 404 Invalid Path`. The normal 404 status line is `HTTP/1.1 404 Not Found` and I think anything else is surprising to people. A simpler way to do this would be to use the Twisted class for 404 which sets the response code, status, etc. and renders a small HTML response indicating the problem:

```
response = resource.NoResource()
return response.render(request)
```

pypingou replied:
having a response body (being JSON since we use JSON everywhere) is helpful as a way to saying what wasn't found, was it the user, a config or just the wrong url?

jeremycline replied:
Okay. The current implementation is returning a content-type of `text/html` rather than `application/json`. Including more detail about what resource is missing sounds good, but right now it's just the HTTP status line repeated.

It might be good to define a subclass of `resource.NoResource` which includes the response template and documents the the JSON schema that is returned to the user.

jeremycline committed:
Add a 404 error page class that handles JSON bodies

By default, this class renders as HTML, but it accepts a ``detail``
parameter which is JSON-serialized and used as the response body.

Signed-off-by: Jeremy Cline <jeremy@jcline.org>

jeremycline commented:
I've rebased and added a class to handle rendering the page with the correct `Content-Type`. For now it doesn't provide a `detail` argument and therefore renders as HTML, but it handles JSON so when we can add that.

skrzepto commented:
the function `def invalid_request(self, request, code=404, reason="Invalid Request"):` should also be removed since `JsonNotFound` handles that. Seems after removal of the function the tests still pass

jeremycline commented:
`invalid_request` has been removed

jeremycline commented:
@skrzepto if it's alright with you, I'd like to merge this separately from #52, and I'll rebase that afterwards. It'll make its diff smaller and easier to review.

skrzepto commented on fmn/sse/sse_webserver.py:
```diff
@@ -29,17 +31,15 @@ class SSEServer(resource.Resource):
-        request = self.add_headers(request)
-        request.write(b"")
         if self.is_valid_path(request.postpath):
+            request = self.add_headers(request)
```
I forgot to write `request.write(b"")` to set the headers to the client

jeremycline replied:
Fixed.

jeremycline commented:
Merged as part of #52

pypingou commented:
Should we delete the branch then?

jeremycline commented:
@pypingou yup, good call. Done.
